projects/certs/X509_certificate_to_der.h

```diff
@@ -10,39 +10,37 @@
 
 namespace x509_certificate {
 
-class CertToDER {
- public:
-  // Encodes |X509_certificate| to DER, returning the encoded bytes in |der_|.
-  std::vector<uint8_t> X509CertificateToDER(
-      const X509Certificate& X509_certificate);
-
- private:
-  // Contains DER encoded X509 Certificate.
-  std::vector<uint8_t> der_;
-
-  // Used to encode PDU's for fields that contain them.
-  asn1_pdu::ASN1PDUToDER pdu_to_der;
-
-  // Used to encode the ASN1 types that appear in X509 Certificates.
-  asn1_universal_types::ASN1UniversalTypesToDER u_types_to_der;
-
-  template <typename T>
-  void Encode(const T& t);
-
-  // Encode(FIELD) DER encodes the field found in X509 Certificates
-  // and writes the results to |der_|.
-  void Encode(const TBSCertificateValue& tbs_certificate);
-  void Encode(const VersionNumber& version);
-  void Encode(const ValidityValue& validity);
-  void Encode(const TimeChoice& val);
-  void Encode(const SubjectPublicKeyInfoValue& subject_public_key_info);
-  void Encode(const AlgorithmIdentifier& algorithm_identifier);
-  void Encode(const asn1_pdu::PDU& pdu);
-  void Encode(const asn1_universal_types::Integer& integer);
-  void Encode(const asn1_universal_types::BitString& bit_string);
-  void Encode(const asn1_universal_types::UTCTime& utc_time);
-  void Encode(const asn1_universal_types::GeneralizedTime& generalized_time);
-};
+// Encodes |X509_certificate| to DER, returning the encoded bytes in |der_|.
+std::vector<uint8_t> X509CertificateToDER(
+    const X509Certificate& X509_certificate);
+
+template <typename T>
+void Encode(const T& t, std::vector<uint8_t>& der);
```

Comment:
For the generic template, we want that to be in the header file, since it needs to be extensible for all types.

This largely works now because you've got everything within the single file (e.g. everything that might call this variant is within the same translation unit), but once you split them out, it's important for the generic function to be in the `.h` so the compiler knows how to generate the function.

```suggestion
template <typename T>
void Encode(const T& t, std::vector<uint8_t>& der) {
  if (t.has_pdu()) {
    return Encode(t.pdu(), der);
  }
  return Encode(t.value(), der);
}
```

(and removing from the `.cc`)